### PR TITLE
Add commodity flow of capacity variables

### PR DESCRIPTION
**Pull request by Jihoon**
This PR cleans up and supersedes #387.

### CODE REVIEW

**Comment by khaeru** [comment]
#452 is modifying the CI setup. Please restart the tests here (manually or via rebase) once that one is merged, hopefully within an hour.

**Review by khaeru — CHANGES_REQUESTED** [review]
I edited the PR description to re-add the checklist that's auto inserted by the template. These items, particularly documentation, are *essential*—even for minor PRs. This one is not minor, because it alters the core model formulation.

At the absolute minimum, when completing the item “Update release notes,” mention the parameters, variables, and equations that are added/affected by the additions here. Then, users might have a thin hope of understanding that changes *have been made*, and might try to puzzle them out or even use them.

Much better would be to update the inline documentation in the GAMS files, or add a new file in docs/, with plain-language description. For a good example, see #190 by @behnam-zakeri: https://github.com/iiasa/message_ix/pull/190/files, in particular
- additions in model_core.gms and parameter_def.gms.
- how the corresponding item in the release notes links to these descriptions.

As a reminder, if time feels tight, there is always the option to task a specific colleague with some or all of this work.

**Comment by khaeru** [comment]
Postponed to v3.4 because:
- PR checklist is incomplete.
- 3 requested reviews not done.
- Requested changes from 1 review not completed.

**Comment by LauWien** [comment]
I've added the WIP tag, since there is still more time/investigation needed to finish up the PR, as discussed in today's MESSAGEix meeting. Due to that and the following
- PR checklist is incomplete,
- 3 requested reviews not done,
- Requested changes from 1 review not completed,

we postpone to v3.5.

File: message_ix/model/MESSAGE/model_core.gms [hunk 1]
```diff
@@ -572,6 +584,47 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+  + SUM( (location,tec)$( inv_tec(tec) AND map_tec(location,tec,year) ),                                               \
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time'
+        output_cap_new(location,tec,year,node,commodity,level,time)                                                    \
+        * CAP_NEW(location,tec,year)                                                                                   \
+* input by all new capacity of technologies located at 'location' taking from 'node' and 'time'
+        - input_cap_new(location,tec,year,node,commodity,level,time)                                                   \
+        * CAP_NEW(location,tec,year) )                                                                                 \
+* commodity input and output associated with retirement of technology capacity (via differentials of capacity of successive periods)
+* for first model period (differential with historical remaining capacity)
+  + SUM( (location,tec,vintage,year2)$( inv_tec(tec) AND map_tec_lifetime(location,tec,vintage,year2)                  \
```
Comment (GamzeUnlu95) [inline message_ix/model/MESSAGE/model_core.gms:598]: If we say the first model year is 700 and the vintage is 690:
According to this equation which is valid only for the first model year, based on the conditions specified here 
year  = 700, year2 = 690. In this case for a technology that has vintage 690 map_tec_lifetime(location,tec,690,690) becomes a condition that needs to be satisfied. And map_tec_lifetime does not have this entry for vintage years. In this case there will be no scrap released for the first model year?

File: message_ix/model/MESSAGE/model_core.gms [hunk 2]
```diff
@@ -572,6 +584,47 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+            AND first_period(year) AND seq_period(year2,year) ),                                                       \
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time' distributed over years of periods
+        output_cap_ret(location,tec,vintage,node,commodity,level,time)                                                 \
+        * ( remaining_capacity(node,tec,vintage,year2) - CAP(location,tec,vintage,year) )                              \
```
Comment (GamzeUnlu95) [inline message_ix/model/MESSAGE/model_core.gms:602]: The remaining capacity seems to be a binary value, does it make sense to perform a subtraction such as 
( remaining_capacity(node,tec,vintage,year2) - CAP(location,tec,vintage,year) )?

File: message_ix/model/MESSAGE/model_core.gms (hunk 1 above)
Comment (Jihoon) [reply message_ix/model/MESSAGE/model_core.gms:598]: Just to add a bit on Gamze's comment above, for the capacity supposed to end its activity in 690 and retire in 700, the historical capacity that existed in 690 cannot be taken into account, because the `map_tec_lifetime` condition does not pass for the period with `year`=700. `map_tec_lifetime(location,tec,vintage,year2)` is only defined for `year2 `>= `fmy`.

File: message_ix/model/MESSAGE/model_core.gms [hunk 3]
```diff
@@ -572,6 +584,47 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+        / duration_period(year)                                                                                        \
+* input by all new capacity of technologies located at 'location' taking from 'node' and 'time' distributed over years of periods
+        - input_cap_ret(location,tec,vintage,node,commodity,level,time)                                                \
+        * ( remaining_capacity(node,tec,vintage,year2) - CAP(location,tec,vintage,year) )                              \
+        / duration_period(year) )                                                                                      \
+* for other model periods (differential with installed capacity of preceding period)
+  + SUM( (location,tec,vintage,year2)$( inv_tec(tec) AND map_tec_lifetime(location,tec,vintage,year2)                  \
+            AND NOT first_period(year) AND seq_period(year2,year) ),                                                   \
```
Comment (Jihoon) [inline message_ix/model/MESSAGE/model_core.gms:610]: Another suspected issue here is:
For the capacity ending its activity in e.g. `year2=750`, this part of the equation brings in the term `CAP(location,tec,vintage,year)` where `year=760`. And it appears that while solving the LP model, a spurious non-zero value is assigned to this CAP of `year=760`, which is after the capacity is retired.

File: message_ix/model/MESSAGE/model_core.gms (hunk 2 above)
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:608]: Here we can use something like: 

* output by all new capacity of technologies located at 'location' sending to 'node' and 'time' distributed over years of periods
```
        output_cap_ret(location,tec,vintage,node,commodity,level,time)                                                 
        ( historical_new_capacity(node,tec,vintage) - CAP(location,tec,vintage,year)/ duration_period(year) )  
```      
                                                                                         
* input by all new capacity of technologies located at 'location' taking from 'node' and 'time' distributed over years of periods
```
        - input_cap_ret(location,tec,vintage,node,commodity,level,time)                                               
        (historical_new_capacity(node,tec,vintage) - CAP(location,tec,vintage,year)/ duration_period(year))) 
```

**Comment by GamzeUnlu95** [comment]
> The issue related to the historical vintage `690` of `coal_ppl` seems to be a result of the missing `technical_lifetime` parameter of that vintage in the Westeros tutorial (see #609).
> 
> The problem of non-zero `CAP` values of technologies after the end of their lifetime is due to the fact that there is no equation that would require `CAP` to be zero after the end of the technology lifetime. The equation `CAPACITY_MAINTENANCE` just requires that `CAP` in a period is lower than in a previous period. This usually does not lead to problems, but in this extension there might be a value in delaying commodity release to better balance a "scrap-type" commodity equation that is set to equal. Usually, there is no cost penalty and also no other benefits as other parameters have not been defined for periods beyond the lifetime.
> 
> To avoid this behavior, we may need to introduce an additional equation of the following kind:
> 
> ```
> CAPACITY_MAINTENANCE2(node,inv_tec,vintage,year)$( map_tec(node,inv_tec,vintage) AND NOT map_tec_lifetime(node,inv_tec,vintage,year) 
> AND NOT first_period(year) AND year_order(vintage) < year_order(year))..
>     CAP(node,inv_tec,vintage,year) =L= 0 ;
> ```

After the addition of this equation, the issues related to CAP remaining after the technology lifetime is solved and scrap is generated correctly in the years that are not the first model period.

File: message_ix/model/MESSAGE/model_core.gms [hunk 4]
```diff
@@ -577,6 +590,47 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
 * export from node and input into technologies located at 'location' taking from 'node' and 'time2' taking from 'time'
         - input(location,tec,vintage,year,mode,node,commodity,level,time2,time)                                        \
         * duration_time_rel(time,time2) * ACT(location,tec,vintage,year,mode,time2) )                                  \
+* +++++
+* commodity input and output associated with construction of new technology capacity (during vintage period)
+  + SUM( (location,tec)$( inv_tec(tec) AND map_tec(location,tec,year) ),                                               \
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time'
+        output_cap_new(location,tec,year,node,commodity,level,time)                                                    \
+        * CAP_NEW(location,tec,year)                                                                                   \
+* input by all new capacity of technologies located at 'location' taking from 'node' and 'time'
```
Comment (GamzeUnlu95) [inline message_ix/model/MESSAGE/model_core.gms:599]: `input_cap_new `parameter works as expected in all model years as tested with a simple Westeros tutorial.

**Comment by volker-krey** [comment]
@GamzeUnlu95, the issue was the use of the set `year2` which only covers years as of the first model year. Replacing it by `year_all2` does the job, I believe.

File: message_ix/model/MESSAGE/model_core.gms [hunk 5]
```diff
@@ -577,6 +590,59 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+        - input_cap_new(location,tec,year,node,commodity,level,time)                                                   \
+        * CAP_NEW(location,tec,year) )                                                                                 \
+* commodity input and output associated with retirement of technology capacity (via differentials of capacity of successive periods)
+* for first model period (differential with historical remaining capacity)
+  + SUM( (location,tec,vintage,year_all2)$( inv_tec(tec) AND map_tec_lifetime_extended(location,tec,vintage,year_all2)                  \
```
Comment (volker-krey) [inline message_ix/model/MESSAGE/model_core.gms:604]: The current formulation still does not treat commodity flows from historical vintages that have lifetimes that are not multiples of period length correctly. Those vintages only create a commodity flow in the first model period, but not in any later periods. In addition, the multiplication with `remaining_capacity_extended` implies that longer lifetime leads to higher flows although the opposite should happen, since longer lifetime means less retired capacity.

Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:604]: The vintage years are still defined in the CAP variable if they are available in any model period. This makes sure that after the first model period all the capacity retirements are handeled correctly. The formualtion for the `remaining_capacity_extended` is revised.

**Comment by GamzeUnlu95** [comment]
@khaeru do you have any idea about the failing tests here: 

1) mypy:  `message_ix/tests/test_feature_vintage_and_active_years.py:29: note: Possible overload `. And two more errors in the same line. The branch is recently rebased and this section is the same as the main branch. 

2) Tutorial test which is testing the notebooks wheter they can be run without errors or not is failing (test_tutorials.py line 110): R_austria and R_austria_load_scenario. (Only in Test / windows-latest-py3.10 and Test / windows-latest-py3.7).  I am able to run both of these in my own computer with this new branch.

**Comment by khaeru** [comment]
> 1. mypy:  `message_ix/tests/test_feature_vintage_and_active_years.py:29: note: Possible overload `. And two more errors in the same line. The branch is recently rebased and this section is the same as the main branch.

This might be due to an updated version of mypy. It sometimes introduces new, more precise type checking, so our existing code that was okay with previous mypy versions starts to fail. I can investigate this one.

> 2. Tutorial test which is testing the notebooks wheter they can be run without errors or not is failing (test_tutorials.py line 110): R_austria and R_austria_load_scenario. (Only in Test / windows-latest-py3.10 and Test / windows-latest-py3.7).  I am able to run both of these in my own computer with this new branch.

This is what's called a "flaky" test: sometimes, randomly, it times out and fails on the GitHub Actions runners. The response is just to re-run that 1 failed job, which I'll do for you now.

**Comment by khaeru** [comment]
> This might be due to an updated version of mypy.

Checking locally:
- mypy 0.961, numpy 1.21.5 —passes
- mypy 0.961, numpy 1.23.0 —shows the same error as the failing lint check, which shows numpy 1.22.x.

So the issue appears to be the newer numpy, not a new mypy. I will make a separate PR, then you can rebase this one.

**Comment by khaeru** [comment]
> I will make a separate PR, then you can rebase this one.

Oliver helpfully reviewed that PR. Rebasing this branch on `main` should make the lint error disappear.

File: message_ix/model/MESSAGE/model_core.gms [hunk 6]
```diff
@@ -577,6 +590,61 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+* for lifetimes out of duraion_period, remaining capacity is different than 1 or 0 in the first model period. 
+  + SUM( (location,tec,vintage,year_all2)$( inv_tec(tec) AND map_tec_lifetime_extended(location,tec,vintage,year_all2)                      \
+           AND first_period(year) AND seq_period(year_all2,year)                                                                             \
+           AND (remaining_capacity_extended(location,tec,vintage,year)< 1) AND (remaining_capacity_extended(location,tec,vintage,year)>0)     \
+),                                                                                                                                             \                                                                                                                           
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time' distributed over years of periods
+        output_cap_ret(location,tec,vintage,node,commodity,level,time)                                                 \
+        *  (historical_new_capacity(node,tec,vintage) * (1 - remaining_capacity_extended(node,tec,vintage,year)))            \
```
Comment (Jihoon) [inline message_ix/model/MESSAGE/model_core.gms:624]: I wasn't entirely sure why it should be `(1 - remaining_capacity_extended)` here. If `remaining_capacity_extended `= 0.3 for example, doesn't it mean the lifetime covers only 30% of the duration of the first period? Then the amount of retiring capacity is also just 30% of the `historical_new_capacity`? Pls correct me if I'm not understanding it..

**Review by Jihoon — COMMENTED** [review]
It is indeed a nice and meaningful improvement. I left some comments, many of them about the documentation. It does look like a huge effort. Thanks a lot!

File: message_ix/model/MESSAGE/model_core.gms (hunk 6 above)
Comment (Jihoon) [reply message_ix/model/MESSAGE/model_core.gms:624]: I can see Gamze's point here. So for the first period if remaining_capacity_extended=0.3, 70% is retiring during the first period. So this seems ok then. But then I wonder how the retirement of that 30% is considered in the second period. The next SUM() block below is just based on CAP differences, which cannot consider this partial retirement?

File: message_ix/model/MESSAGE/parameter_def.gms [hunk 7]
```diff
@@ -211,6 +211,41 @@ Parameter
...
+*    * - input_cap_ret [#tecvintage]_
+*      - ``node_loc`` | ``tec`` | ``year_vtg`` |
+*        ``node_origin`` | ``commodity`` | ``level`` | ``time_origin``
+*      - relative share of input per unit of capacity retired
+*    * - output_cap_ret [#tecvintage]_
+*      - ``node_loc`` | ``tec`` | ``year_vtg`` |
+*        ``node_dest`` | ``commodity`` | ``level`` | ``time_dest``
+*      - relative share of output per unit of capacity retired
+*
 * .. [#tecvintage] Fixed and variable cost parameters and technical specifications are indexed over both
```
[the remark this answers is not in the document]
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/parameter_def.gms:249]: I added a line: Material flows can also vary based on the vintage and active years. Did you have more explanation in mind?

File: message_ix/model/MESSAGE/model_core.gms [hunk 8]
```diff
@@ -560,6 +561,18 @@ RESOURCE_HORIZON(node,commodity,grade)$( SUM(year$map_resource(node,commodity,gr
...
 *     - \sum_{\substack{n^L,t,m,h^A \\ y^V \leq y}} input_{n^L,t,y^V,y,m,n,c,l,h^A,h}
 *         \cdot duration\_time\_rel_{h,h^A} \cdot ACT_{n^L,t,m,y,h^A} & \\
+*     + \sum_{\substack{n^L,t}} output\_cap\_new_{n^L,t,y,n,c,l,h}
+*         \cdot CAP\_NEW_{n^L,t,y} & \\
+*     - \sum_{\substack{n^L,t}} input\_cap\_new_{n^L,t,y,n,c,l,h}
+*         \cdot CAP\_NEW_{n^L,t,y} & \\
+*     + \sum_{\substack{n^L,t \\ y^V \leq y}} output\_cap\_ret_{n^L,t,y^V,n,c,l,h}
+*         \cdot ( CAP_{n^L,t,y^V,y-1} - CAP_{n^L,t,y^V,y} ) \cdot \frac{duration\_period_y}{duration\_period_CAP_y^V} & \\
+*     - \sum_{\substack{n^L,t \\ y^V \leq y}} input\_cap\_ret_{n^L,t,y^V,n,c,l,h}
+*         \cdot ( CAP_{n^L,t,y^V,y-1} - CAP_{n^L,t,y^V,y} ) \cdot \frac{duration\_period_y}{duration\_period_CAP_y^V} & \\
```
[the remark this answers is not in the document]
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:571]: I made the same change as above.

File: message_ix/model/MESSAGE/model_core.gms [hunk 9]
```diff
@@ -560,6 +561,18 @@ RESOURCE_HORIZON(node,commodity,grade)$( SUM(year$map_resource(node,commodity,gr
 *         \cdot duration\_time\_rel_{h,h^A} \cdot ACT_{n^L,t,y^V,y,m,h^A} & \\
 *     - \sum_{\substack{n^L,t,m,h^A \\ y^V \leq y}} input_{n^L,t,y^V,y,m,n,c,l,h^A,h}
 *         \cdot duration\_time\_rel_{h,h^A} \cdot ACT_{n^L,t,m,y,h^A} & \\
+*     + \sum_{\substack{n^L,t}} output\_cap\_new_{n^L,t,y,n,c,l,h}
```
[the remark this answers is not in the document]
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:564]: I tried to align them all to the right side but the lines with fewer entries constantly shift to the left even when I try to combine two short lines.. So if you know how to do that, please make the change. 

About the first-year formulation, in the latest version, there are now two different formulations for the first year also based on a condition of remaning_capacity. So I was not sure whether to include all those special conditions in the main formulation as it might look too crowded. Maybe it is better to describe the differences in a paragraph and direct the reader to the model_core.py file if he/she wants to see more details?

File: message_ix/model/MESSAGE/model_core.gms [hunk 10]
```diff
@@ -752,6 +820,26 @@ CAPACITY_MAINTENANCE(node,inv_tec,vintage,year)$( map_tec_lifetime(node,inv_tec,
...
+* .. _equation_end_of_lifetime_capacity:
+*
+* Equation END_OF_LIFETIME_CAPACITY
+* """""""""""""""""""""""""""""""""
+* This constraint ensures that the capacity is not preserved after the technical lifetime of a technology.
+*
+*   .. math::
+*      CAP_{n,t,y^V,y} =
+*          0 \\
+*      \quad & \text{if } y > y^V \text{ and } y^V  > 'first\_period' \text{ and } |y| - |y^V| >= technical\_lifetime_{n,t,y^V}
```
[the remark this answers is not in the document]
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:833]: Yes, I deleted all of them.

File: message_ix/model/MESSAGE/model_core.gms [hunk 11]
```diff
@@ -577,6 +590,61 @@ $macro COMMODITY_BALANCE(node,commodity,level,year,time) (
...
+* output by all new capacity of technologies located at 'location' sending to 'node' and 'time' distributed over years of periods
+* Based on the condition that remaining capacity in the first year is 0. 
+        output_cap_ret(location,tec,vintage,node,commodity,level,time)                                                 \
+        *  (historical_new_capacity(node,tec,vintage) * remaining_capacity_extended(node,tec,vintage,year_all2))            \
+* input by all new capacity of technologies located at 'location' taking from 'node' and 'time' distributed over years of periods
+        - input_cap_ret(location,tec,vintage,node,commodity,level,time)                                                \
+        *  (historical_new_capacity(node,tec,vintage) * remaining_capacity_extended(node,tec,vintage,year_all2))) \      
+* commodity input and output associated with retirement of technology capacity (via differentials of capacity of successive periods)
+* for first model period (differential with historical remaining capacity)
+* for lifetimes out of duraion_period, remaining capacity is different than 1 or 0 in the first model period.
```
[the remark this answers is not in the document]
Comment (GamzeUnlu95) [reply message_ix/model/MESSAGE/model_core.gms:617]: Yes, I meant that. I changed the wording to be more clear. If `duration_period` is 10 and the technical lifetime is 25 for example.

**Comment by glatterf42** [comment]
Agreed with @GamzeUnlu95 to move this to 3.8.

**Comment by glatterf42** [comment]
@GamzeUnlu95, @Jihoon, @macflo8: what is the status of this PR? Are you currently using it? What needs to be done to finish it off? Do you think it is realistic to do so by the end of this year?

**Comment by glatterf42** [comment]
I'm going to assume that this won't be ready this week, so we will postpone it to the 3.9 milestone. However, I'd be happy to help here, so please let me know what this PR requires to be ready to merge.

**Comment by macflo8** [comment]
@glatterf42 Sorry for the delayed response. This branch contains additions to the MESSAGEix GAMS code. The additions are run by default and add ~10 minutes to the total runtime per MESSAGE iteration. Merging this PR in its current state would therefore negatively affect the runtime of all MESSAGEix users. Thus, we have to postpone the merge for now. We are planning revisit this PR next year and add a mechanism to run the additional GAMS code not by default but only optionally.

**Comment by glatterf42** [comment]
Thanks for the update! Unfortunately, I probably won't be able to help writing GAMS code, but moving to the next milestone then seems like the right call :)

**Comment by glatterf42** [comment]
@khaeru suggested (when this PR was discussed in today's monthly meeting), that @daymontas1 recently improved the MACRO code so that regions can solve simultaneously and that similar improvements may be possible here.

**Comment by khaeru** [comment]
> @khaeru suggested (when this PR was discussed in today's monthly meeting), that @daymontas1 recently improved the MACRO code so that regions can solve simultaneously and that similar improvements may be possible here.

To expand a bit on that suggestion:

- Here's [`message_ix.Scenario.solve()`](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/core.py#L688-L707). It takes some `solve_options` and generic `kwargs`. Then it calls…
- [`ixmp.Scenario.solve()`](https://github.com/iiasa/ixmp/blob/ac0bba16724ce75caba4adb970d9d2d1f6850a20/ixmp/core/scenario.py#L823-L920). This method…
- First, creates, a `ixmp.model.base.Model` (or a subclass), thus calling its [`__init__()`](https://github.com/iiasa/ixmp/blob/ac0bba16724ce75caba4adb970d9d2d1f6850a20/ixmp/model/base.py#L23-L35) method.
- Then, Scenario.solve() calls [`Model.run()`](https://github.com/iiasa/ixmp/blob/ac0bba16724ce75caba4adb970d9d2d1f6850a20/ixmp/model/base.py#L187-L200).
- In the case of `message_ix.Scenario`, the specific class used is one of 3 subclasses of [`ixmp.model.gams.GAMSModel`](https://github.com/iiasa/ixmp/blob/ac0bba16724ce75caba4adb970d9d2d1f6850a20/ixmp/model/gams.py#L97) (provides generic features for invoking GAMS) → [`message_ix.models.GAMSModel`](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L148-L149) (provides specific features needed by `message_ix`) → one of [`MESSAGE`](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L266-L267), [`MACRO`](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L901-L902), or [`MESSAGE_MACRO`](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L985-L986).
- Each of these classes and methods has the option to handle the arguments from `kwargs` and `solve_options` as they go through; to pass them on; to do various things; to raise exceptions; etc. For example:
  - ixmp.model.base.GAMSModel handles [these arguments](https://github.com/iiasa/ixmp/blob/ac0bba16724ce75caba4adb970d9d2d1f6850a20/ixmp/model/gams.py#L167-L177).
  - message_ix.models.GAMSModel deals with [CPLEX options](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L180-L187) from the user's config or arguments. It also [handles the cplex.opt file](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L192-L193).
  - message_ix.models.MESSAGE_MACRO handles the [meta-parameters that control the MESSAGE-and-MACRO iteration](https://github.com/iiasa/message_ix/blob/7dce4617b2ca6ca17a03cbe20fcc8e57fd56fbae/message_ix/models.py#L992-L1000).

So among these, there are many opportunities to do things like (a) pass a GAMS parameter that would affect conditional inclusion of other GAMS files (that would be the approach in this PR: unless particular features are switched on, *do not* include the .gms files that enable that feature but lead to a slow compile) or (b) set a variable or something else that the GAMS code would respond to (this could be the approach in #808 to choose sequential or parallel solve of MACRO).

*Where* precisely to handle/set defaults/etc. for new options/arguments is easily determined by their scope of use:
- Any GAMS model usable with ixmp (or ixmp4)-stored data? → ixmp.model.gams.GAMSModel.
- All of the models shipped with message_ix? → message_ix.models.GAMSModel.
- Combinations involving MESSAGE, e.g. MESSAGE and MESSAGE_MACRO? → message_ix.models.MESSAGE.
- Only MACRO? → message_ix.models.MACRO.

I think by following the structure and example of existing code it is fully feasible to integrate this and the other PR.

**Comment by glatterf42** [comment]
@Jihoon @GamzeUnlu95 @macflo8 Is there any update on this PR?
Since we probably want to publish a 3.9 release still this week, I'll postpone this to the 3.10 milestone.

**Comment by glatterf42** [comment]
@GamzeUnlu95 and me discussed bilaterally that the status of this PR remains unchanged: the additional run time for all message runs needs to be addressed before the PR can get merged. It does not look like @GamzeUnlu95 will have enough time to do so before leaving IIASA. This PR might therefore have to wait for @GamzeUnlu95's successor and will likely be stale in the meantime -- unless it's specifically assigned to someone else.

**Comment by yiyi1991** [comment]
Thanks for the suggestions. 

So the next thing that will be added to this PR is a solve option "material_stock" in the `class MESSAGE(GAMSModel)` in message_ix.models.
```python
material_stock = str(kwargs.pop("material_stock"))
```
And corresponding updates in message_ix.model.MESSAGE.model_core.gams, and the like. 
```GAMS
$IFTHEN %MATERIAL_STOCK% == "1"
```
Finally adding solves with and without this solve option (printing solving time) in the test. 

Let me know if this is a wrong direction.

**Comment by khaeru** [comment]
@yiyi1991 that's indeed the correct direction.

For reference, you can look at #808, which I updated to serve as an example for both this and for @ywpratama in #908. Focus on MACRO_CONCURRENT and see how it is set in the tests, via the .models classes, and in the GAMS source. Please ask any questions or if you like make a first cut at implementing something similar on this branch and then ask for (non-final) review.

**Comment by yiyi1991** [comment]
(Just to refresh our memory)
- Why is this PR not merged yet? 
The equations are ready but solving with capacity commodities takes a longer time. 
- What is needed? 
A switch that can solve the model with or without capacity commodities.

The switch is now added. 

A scenario `scen` can be solved in this way:
`scen.solve(gams_args=["--cap_comm = 1"])`

- What needs to be checked? 

| scenario | an R12 scenario with par input_cap_new | an R12 scenario without par input_cap_new |
|----------|-----------------------|---------------------------|
| solve by message_ix main | A (took 03:32) | B (took 03:42) |
| solve by message_ix material_stock | C (took 14:16) | D (code 2)|
| solve by cap_comm = 0 | E (took 03:49) | F (took 03:42)|
| solve by cap_comm = 1 | G (took 14:29)| H (code 2)|

- Did the following scenarios generate some results? 
A = E
C = G
B = F
Yes, all checked. 

**- Did it shorten the solving time?** 
Yes, the solving time of E is shorter than C.

P.s., the emission budgets of the two scenarios are similar but not exactly the same, which may affect the solving time a bit; the solved A and E show basically identical ACT and CAP, but slight differences in `PRICE_EMISSION` and `PRICE_COMMODITY` (e.g., 0.0026% diff in `PRICE_EMISSION` TCE).

**Comment by khaeru** [comment]
I will step in here to:
- Rebase on latest `main`.
- Resolve the conflicts.
- Resolve ~20 failing tests.
- Align the switch with the code and docs style used in #808.
- Edit a few commit messages to align with our code style.

**Comment by khaeru** [comment]
After having looked at this for most of day, I'm still a bit confused by one of the clauses/sums added to the new COMMODITY_BALANCE_AUX equation:
https://github.com/iiasa/message_ix/blob/3422870090503738a36b944ceb78a386fa9e3fab/message_ix/model/MESSAGE/model_core.gms#L669-L692

…in particular:
- The way `map_cap_ret_hist_1` is constructed ([here](https://github.com/iiasa/message_ix/blob/3422870090503738a36b944ceb78a386fa9e3fab/message_ix/model/MESSAGE/scaling_investment_costs.gms#L189-L197)):
  - The last dimension/index `year` can _only_ be `first_period` (call that $y_0$), for instance 2020; 
  - The second-last dimension/index `year_all2` can _only_ be the period immediately preceding $y_0$; call this $y_{-1}$, for instance 2015.
- Thus the clause labelled (4) appears to mean:
  - For each (node, tec, year_vtg=`vintage`)…
  - Take the remaining_capacity **in the period** $y_{-1}$.
  - Multiply by the amount of `historical_new_capacity` from the respective year_vtg.
  - Multiply by the corresponding `{in,out}put_cap_ret`.
  - Sum over all (node, tec, year_vtg).

What confuses me is why this is counted towards the COMMODITY_BALANCE in `year`/$y_0$ when the retirement in question is happening in $y_{-1}$.

I'd like to understand this before I go on to add the docs needed to complete the PR.

**Comment by khaeru** [comment]
Postponed to 3.12 to allow time for a clear resolution of the above point, documentation, and tests.

**Comment by khaeru** [comment]
I'll rebase this branch to resolve conflicts.

**Comment by khaeru** [comment]
Another test failure:
```
FAILED message_ix/tests/test_macro.py::test_calc_price_zero[jdbc] - AssertionError: Regex pattern did not match.
```
This boils down to the following: on `main`, the `PRICE_COMMODITY` value in the solved Westeros scenario look like:
```
    node    commodity      level  year  time            lvl  mrg
Westeros  electricity  secondary   710  year  4.940656e-324  0.0
Westeros  electricity  secondary   720  year  4.940656e-324  0.0
Westeros  electricity      final   700  year  4.940656e-324  0.0
Westeros        light     useful   700  year  4.940656e-324  0.0
Westeros        light     useful   710  year  4.940656e-324  0.0
Westeros        light     useful   720  year   5.000000e+00  0.0
```
On this branch, we instead see:
```
    node    commodity      level  year  time            lvl  mrg
Westeros  electricity  secondary   700  year  4.940656e-324  0.0
Westeros  electricity  secondary   710  year  4.940656e-324  0.0
Westeros  electricity      final   700  year  4.940656e-324  0.0
Westeros  electricity      final   710  year  4.940656e-324  0.0
Westeros  electricity      final   720  year  4.940656e-324  0.0
Westeros        light     useful   720  year   5.000000e+00  0.0
```

The missing values for (light, useful, 700), (light, useful, 710) cause a preliminary calculation to fail, before reaching the main focus of the test.

Some thing I observe:
- In **neither case**—not on `main` or this branch—does PRICE_COMMODITY include all of the "active" combinations of (node, commodity, level, year, time)—that is, the indices for which there are non-zero in- or outflows of commodities.
- The descriptive text for PRICE_COMMODITY says

  > commodity price (derived from marginals of COMMODITY_BALANCE constraint)

  - This was set in a commit from 2018: https://github.com/iiasa/message_ix/blame/9fb0b97867f9234d98b4e43e158d7af9b30c42f4/message_ix/model/MESSAGE/model_core.gms#L148
  - That actually precedes commit 753e911bc0f92fc19da38e4961cdc40ee7b239b5 that changed COMMODITY_BALANCE from an equation to a macro (the current PR is changing it back).

- It is defined here: https://github.com/iiasa/message_ix/blob/9fb0b97867f9234d98b4e43e158d7af9b30c42f4/message_ix/model/MESSAGE/model_solve.gms#L52-L55
- I note that the new COMMODITY_BALANCE_AUX equation **does** have marginal values present for all "active" (n, c, l, y, h) combinations. I see values like:
```
$ gdxdump X.gdx

Equation COMMODITY_BALANCE_AUX(node,commodity,level,year_all,time) Auxiliary equation for calculation of commodity balance /
'Westeros'.'light'.'useful'.'720'.'year'.M -8.93317663555801,

Equation COMMODITY_BALANCE_GT(*,*,*,*,*) commodity supply greater than or equal demand /
'Westeros'.'light'.'useful'.'720'.'year'.M 8.93317663555801,

Equation COMMODITY_BALANCE_LT(*,*,*,*,*) commodity supply lower than or equal demand / /;
```
  - In other words, the value of COMMODITY_BALANCE_AUX.M appears to be the negative of (COMMODITY_BALANCE_GT.M + COMMODITY_BALANCE_LT.M), the quantity 

Based on these observations, I will try to revise PRICE_COMMODITY to refer to COMMODITY_BALANCE_AUX instead of COMMODITY_BALANCE_{GT,LT}.
